File: src/main/java/com/lbx/demo/model/TitleCrew.java
package com.lbx.demo.model;

import lombok.*;

import java.util.*;
import java.util.stream.Collectors;

/**
 * 2. Return all the titles in which both director and writer are the same person and he/she is still alive
 */
@AllArgsConstructor
@NoArgsConstructor
@ToString
@Getter
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class TitleCrew implements FieldSettable {
    private String titleId;
    private final List<String> directors = new ArrayList<>();
    private final List<String> writers = new ArrayList<>();

    public Set<String> sameDirectorAndWriter() {
        return directors.stream().filter(writers::contains).collect(Collectors.toSet());
    }

    @Override
    public void setField(String name, String value) {
        switch (name) {
            case "tconst" -> titleId = value;
            case "directors" -> directors.addAll(Arrays.asList(value.split(",")));
            case "writers" -> writers.addAll(Arrays.asList(value.split(",")));
        }
    }

    public TitleCrew addDirector(String nameId) {
        directors.add(nameId);
        return this;
    }

    public TitleCrew addWriter(String nameId) {
        writers.add(nameId);
        return this;
    }
}

File: src/main/rust/src/models/title_crew.rs
use std::collections::HashSet;

use crate::models::mapper::FieldSettable;

#[derive(Debug, Default, Clone)]
pub struct TitleCrew {
    pub title_id: String,
    pub directors: HashSet<String>,
    pub writers: HashSet<String>,
}

impl FieldSettable for TitleCrew {
    fn new() -> Self {
        TitleCrew::default()
    }
    fn set_field(&mut self, name: &str, value: &str) {
        match name {
            "tconst" => self.title_id = value.to_string(),
            "directors" => value.trim().split(',').for_each(|g| {
                self.directors.insert(g.trim().to_string());
            }),
            "writers" => value.trim().split(',').for_each(|g| {
                self.writers.insert(g.trim().to_string());
            }),
            _ => {
                panic!("no such column![{name}]")
            }
        }
    }
}

impl TitleCrew {
    pub fn same_director_and_writer(&self) -> HashSet<String> {
        self.directors
            .intersection(&self.writers)
            .map(|s| s.clone())
            .collect()
    }
}
